src/vmspawn/vmspawn.c

```diff
@@ -605,71 +605,100 @@ static int open_vsock(void) {
         return TAKE_FD(vsock_fd);
 }
 
-static int vmspawn_dispatch_notify_fd(sd_event_source *source, int fd, uint32_t revents, void *userdata) {
-        char buf[NOTIFY_BUFFER_MAX+1];
-        const char *p = NULL;
-        struct iovec iovec = {
-                .iov_base = buf,
-                .iov_len = sizeof(buf)-1,
-        };
-        struct msghdr msghdr = {
-                .msg_iov = &iovec,
-                .msg_iovlen = 1,
-        };
-        ssize_t n;
-        _cleanup_strv_free_ char **tags = NULL;
-        int r, *exit_status = ASSERT_PTR(userdata);
+typedef struct NotifyConnectionData {
+        char buffer[NOTIFY_BUFFER_MAX+1];
+        size_t full;
+        int *exit_status;
+} NotifyConnectionData;
 
-        n = recvmsg_safe(fd, &msghdr, MSG_DONTWAIT);
-        if (ERRNO_IS_NEG_TRANSIENT(n))
-                return 0;
-        if (n == -EXFULL) {
-                log_warning_errno(n, "Got message with truncated control data, ignoring: %m");
-                return 0;
-        }
-        if (n < 0)
-                return log_warning_errno(n, "Couldn't read notification socket: %m");
+static int read_vsock_notify(NotifyConnectionData *d, int fd) {
+        int r;
 
-        if ((size_t) n >= sizeof(buf)) {
-                log_warning("Received notify message exceeded maximum size. Ignoring.");
-                return 0;
+        assert(d);
+        assert(fd >= 0);
+
+        for (;;) {
+                assert(d->full < sizeof(d->buffer));
+
+                ssize_t n = read(fd, d->buffer + d->full, sizeof(d->buffer) - d->full);
+                if (n < 0) {
+                        if (ERRNO_IS_TRANSIENT(errno))
+                                return 0;
+
+                        return log_error_errno(errno, "Failed to read notification message: %m");
+                }
+                if (n == 0) /* We hit EOF! Let's parse this */
+                        break;
+
+                d->full += n;
+
+                if (d->full >= sizeof(d->buffer))
+                        return log_error_errno(SYNTHETIC_ERRNO(EBADMSG), "Received notify message exceeded maximum size.");
```

Comment:
Seems a bit extreme to stop reading completely if we get one message that's too big.